### ISSUE TITLE
ConPTY: Entire line is reprinted for every prompt character after a space

### ISSUE THREAD

**Issue by Tyriar** [microsoft/terminal#1647]
Related: https://github.com/microsoft/terminal/issues/1646

# Environment

```none
Windows build number: Microsoft Windows [Version 10.0.18922.1000]
vscode 1.36.0 insiders
```

# Steps to reproduce

Repros in both cmd and python

1. Type a
2. Press space
3. Repeat

# Expected behavior

Only a and space should be sent from conpty until the end of the line is reached

# Actual behavior

```
a[0K
[27G
[?25l
C:\Users\daimms.REDMOND>a a[0K[?25h
[29G
[?25l
C:\Users\daimms.REDMOND>a a a[0K[?25h
[31G
[?25l
C:\Users\daimms.REDMOND>a a a a[0K[?25h
[33G
[?25l
C:\Users\daimms.REDMOND>a a a a a[0K[?25h
[35G
[?25l
C:\Users\daimms.REDMOND>a a a a a a[0K[?25h
[37G
[?25l
C:\Users\daimms.REDMOND>a a a a a a a[0K[?25h
[39G
[?25l
C:\Users\daimms.REDMOND>a a a a a a a a[0K[?25h
[41G
```

**Comment by zadjii-msft** [comment]
Both of these are _horrific_ and I hope we can get to them soon. 

Out of curiosity, does it repro with powershell or wsl? Might be helpful to know if this is related to how `COOKED_READ` handles the prompt line, or if it's for any scenario like this. (powershell and wsl won't be using `COOKED_READ`, while `cmd` will)

**Comment by Tyriar** [comment]
pwsh output:

```
a
[m
a[m 
[25l[7;29Ha[m a[?25h
[25l[7;29Ha[m a [?25h
[25l[7;29Ha[m a a[?25h
[25l[7;29Ha[m a a [?25h
[25l[7;29Ha[m a a a[?25h
```

wsl output:

```
a
 
a
 
a
 
a
```

**Comment by zadjii-msft** [comment]
That at least makes me feel _better_. Looks like it is COOKED_READ related, and if I had to guess, PSReadline redraws the prompt on every character, so that might be out of our control.

**Comment by Tyriar** [comment]
@zadjii-msft hmm... On PSReadline, I'm beginning to think terminals that want to support screen readers well will need to have some buffer diffing as well to avoid reading out prompts when they're reprinted, fish has some fancy auto-complete for example that would probably be awful for screen readers for example and I think zsh might do similar for character attributes.

But this is still a valid report right, cmd.exe shouldn't be reprinting?

**Comment by zadjii-msft** [comment]
Oh yea it's definitely still valid. Now I know where to look :)
<hr>
IIRC the UIA stuff we have in conhost today chokes on things like PSReadline too - taking a larger diff than change-by-change is a good idea, but I'm not super familiar with that code. I doubt it'd be possible with the architecture we currently have. @adiviness can correct me if I'm wrong, and @carlos-zamora may be interested in this side-bar.

**Comment by zadjii-msft** [comment]
@Tyriar Are you still seeing this? When I try booting up python in conpty and following the repro of `a a a a a`, I get the following:

```
^[]0;C:\Users\migrie\AppData\Local\Temp\OpenConsole\15550\conhost.exeSPC-SPCpython^G
PythonSPC3.7.3SPC(v3.7.3:ef4ec6ed12,SPCMarSPC25SPC2019,SPC22:22:05)SPC[MSCSPCv.1916SPC64SPCbitSPC(AMD64)]SPConSPCwin32^M^J
TypeSPC"help",SPC"copyright",SPC"credits"SPCorSPC"license"SPCforSPCmoreSPCinformation.^M^J>>>SPC^[[66X
Input "a" [1]
a
Input "SPC" [1]
SPC
Input "a" [1]
a
Input "SPC" [1]
SPC
Input "a" [1]
a
Input "SPC" [1]
SPC
```

(where `SPC` is a space)

Similarly, I'm only ever seeing `a`s and ` `s when I do the same thing in `cmd`.

There were some optimizations that went into conpty around April, but it looks to me like those actually made it to master around 18883.1000.rsmaster.190417-1356. The task/bug/pr I'm looking at are MSFT:20297559, MSFT:21091917, and MSFT:3121537.

**Comment by Tyriar** [comment]
Fixed in 10.0.18970.1005:

![image](https://user-images.githubusercontent.com/2193314/64541554-e8147580-d2d6-11e9-8a62-96afe03ac148.png)
